extlib/millhone/src/cmd/analyze_jars.rs

```diff
@@ -0,0 +1,204 @@
+    let jar_analysis = JarAnalysis {
+        discovered_jars: jars_in_container(opts.image_tar_file())
+            .with_context(|| format!("Analyze Jar Archive {:?}", tar_filename))?,
+    };
+
+    serde_json::to_writer(std::io::stdout(), &jar_analysis).context("Serialize Results")
+}
+
+/// Extracts the container (saved via `docker save`) and finds JAR files inside any layer.
+/// For each one found, fingerprints it and reports all those fingerprints along with their
+/// layer and path.
+#[tracing::instrument]
+fn jars_in_container(image_path: &PathBuf) -> Result<HashMap<PathBuf, Vec<DiscoveredJar>>, Error> {
+    // Visit each layer and fingerprint the JARs within.
+    info!("inspecting container");
+    // May have to make two copies of this Archive, idk if it can be iterated twice.
+    let layers = list_container_layers(image_path)?;
+    let mut discoveries = HashMap::new();
+
+    let mut image_archive = unpack(image_path)?;
+    for entry in wrap_tar_err!("Iterate entries", image_archive.entries())? {
+        let entry = wrap_tar_err!("Read entry", entry)?;
+        let path = wrap_tar_err!("Read path", entry.path())?;
+        if !layers.contains(path.as_ref()) {
+            debug!(?path, "skipped: not a layer file");
+            continue;
+        }
+
+        let layer = path.to_path_buf();
+        // Layers should have a form like blob
+        let layer_discoveries = jars_in_layer(entry)?;
+        discoveries.insert(layer, layer_discoveries);
+    }
+
+    Ok(discoveries)
+}
+
+/// Open and unpack a file and put it into a tar.
+/// This is done repeatedly because `entries()` can only be read once from an Archive.
+#[tracing::instrument]
+fn unpack(path: &PathBuf) -> Result<Archive<File>, Error> {
+    let file = File::open(path).map_err(Error::OpenFile)?;
+    Ok(tar::Archive::new(file))
+}
+
+#[tracing::instrument(skip(entry))]
+fn jars_in_layer(entry: Entry<'_, impl Read>) -> Result<Vec<DiscoveredJar>, Error> {
+    let mut discoveries = Vec::new();
+
+    let mut entry_archive = tar::Archive::new(entry);
+    for entry in wrap_tar_err!("list entries in layer", entry_archive.entries())? {
+        let entry = wrap_tar_err!("read entry", entry)?;
+        let path = wrap_tar_err!("read path", entry.path())?;
+        if !path.to_string_lossy().ends_with(".jar") {
+            debug!(?path, "skipped: not a jar file");
+            continue;
+        }
+
+        let path = path.to_path_buf();
+        let res = info_span!("jar", ?path).in_scope(|| {
+            debug!("fingerprinting");
+            let entry = buffer(entry)?;
+
+            match Combined::from_buffer(entry) {
+                Ok(fingerprint) => {
+                    discoveries.push(DiscoveredJar { fingerprint, path });
+                    Ok(())
+                }
+                Err(e) => Err(Error::GenerateFingerprint(e)),
+            }
+        });
+
+        if let Err(e) = res {
+            warn!("Fingerprinting: {:?}", e);
+        }
```

Comment:
[nit] emitting the `warn` where it is omits the path from the context; prefer emitting inline unless that was desired:
```suggestion
        let path = path.to_path_buf();
        info_span!("jar", ?path).in_scope(|| {
            debug!("fingerprinting");
            let entry = buffer(entry)?;

            match Combined::from_buffer(entry) {
                Ok(fingerprint) => discoveries.push(DiscoveredJar { fingerprint, path })
                Err(e) => warn!("failed to fingerprint: {e:?}"),
            }
        });
```